### ISSUE TITLE
Running exwm within a daemon server started outside of X

### ISSUE THREAD

**Issue by akater** [ch11ng/exwm#661]
I start emacs daemon server shortly after system boot. Even with `-f exwm-enable`, it doesn't seem to work. Worse still: sometimes with exwm bits loaded in my `init.el`, I can't create any frame whatsoever, including terminal frames, to which only remedy is to edit `init.el` and restart the daemon.

However, it doesn't look totally hopeless. I managed to create a GUI frame for such a daemon once, and it did list X windows as buffers but it was terribly buggy: workspaces didn't work, workspace-related keybindings were undefined, and after I killed GUI frame and tried to switch to a EXWM buffer in tty out of curiosity, I got an error, Emacs froze and even some other ttys exhibited buggy behaviour then.

Overall there was a lot of weirdness, not necessarily easy to reproduce. If needed, I could compile a proper account of what I did and what followed, but only if such a feature as running exwm within a daemon server started outside of X stands a chance to be supported.

For clarity: I always start emacsclient in a tty session before ever running xinit, so e.g. Ctrl+Alt+F1 invoked in X takes me to a nice Emacs environment. For quite some time I was satisfied with running separate daemons for tty use and for X use; it helps when GUI Emacs becomes unresponsive, for example. Nevertheless, I'd like to be able to run a single daemon server instance. Obvious benefits include shared buffers with no need for explicit reverts, and there are some obscure benefits as well. Overall, this feature would provide a more integrated Emacs environment. And maybe it's already there, and I'm just not setting it up right.

I'm using Emacs 27 live master on Gentoo. Running exwm in its own emacs instance set up in `.xinitrc` causes no issues. For completeness, I note that Gentoo does install some additional daemon-related stuff into site-lisp.

**Comment by ch11ng** [comment]
Such use case is definitely supported.  Not sure about your exact settings but please make sure you are using a non-GTK build of Emacs for this purpose.  The Lucid build works just well as I tried.  Please report back if you still get the problem.

**Comment by akater** [comment]
On October 2, 2019 12:59:08 PM UTC, Chris Feng <notifications@github.com> wrote:
>Such use case is definitely supported.  Not sure about your exact
>settings but please make sure you are using a non-GTK build of Emacs
>for this purpose.  The Lucid build works just well as I tried.  Please
>report back if you still get the problem.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/ch11ng/exwm/issues/661#issuecomment-537480439

My Emacs is (and has long been) compiled `--with-x-toolkit=no`.

Great to hear that the feature is supported. My monitor is on its deathbed; a more thorough report is thus delayed.

**Comment by akater** [comment]
medranocalvo <notifications@github.com> writes:

> Please, report the value of `system-configuration-features`.

```
"XPM JPEG TIFF GIF PNG RSVG CAIRO IMAGEMAGICK GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ ZLIB OLDXMENU X11 XDBE XIM THREADS PDUMPER GMP"
```

**Comment by ch11ng** [comment]
@akater So you are still experiencing this problem?

**Comment by akater** [comment]
@ch11ng I've performed more detailed tests. exwm starts fine with my everyday config, with two caveats:
- xinit is started with `exec emacsclient --eval "(exwm-init)" -cs emacs-akater-daemon -F '((fullscreen . fullboth))'`
- everything related to `smart-mode-line` package is removed from emacs init.

Frame parameters (-F) option in emacsclient is added for aesthetics and is likely non-essential
Socket name (-s) option in emacsclient is added for consistency and might be non-essential

Evaluating `(exwm-init)` turns out to be essential for me. (It can be evaluated manually of course.) I'm using `use-package` extensively but the corresponding exwm form has `:demand t` option in it. For completeness: it also has `:after (server xcb)`.

If I leave smart-mode-line section in emacs init, xinit quits shortly after it is started. This never was a problem with standalone emacs. I also have some switches to dynamically alter modeline colors (and theme in general) depending on whether the frame is graphical or not, these have been turned off as well, and I mention this for completeness.

**Comment by akater** [comment]
Now it looks like emacsclient starts fine without exwm-init. I'm sorry for the confusion; I am very confused myself. Some issues are certainly not easily reproducible. For example, exwm-randr sometimes does not work properly, and sometimes it does.

However, there is one persistent issue: if I leave exwm with `C-x C-c`, I can't then create a new frame with `emacsclient -ts emacs-akater-daemon` because the socket can't be found. This never happens with tty frames which I also leave with `C-x C-c`. This might be a generic Emacs issue. I'm afraid I'm done with tests for now, as it takes quite a lot of time.

To sum it up, `smart-mode-line` was likely the core reason behind the issue. It is known to cause problems, e.g. it used to freeze Emacs when used in conjunction with term or ansi-term. https://github.com/Malabarba/smart-mode-line/issues/198

**Comment by ch11ng** [comment]
> Now it looks like emacsclient starts fine without exwm-init.

This works if you already have `(exwm-enable)`.

> exwm-randr sometimes does not work properly, and sometimes it does.

Do you mean with the same configurations? There're something to check when it doesn't work:
* Run `xrandr` and see if the output matches what you configured in `exwm-randr-screen-change-hook`.
* Otherwise, call <kbd>M-x exwm-randr-refresh RET</kbd> and see if the geometries of workspaces are correctly updated.  If not, perhaps there's something wrong with `exwm-randr-workspace-monitor-plist`.

> if I leave exwm with `C-x C-c`, I can't then create a new frame with `emacsclient -ts emacs-akater-daemon` because the socket can't be found.

Do you mean the daemon exited? Otherwise we'll have to find out who stopped the server; EXWM only stops the one it creates internally.

**Comment by akater** [comment]
>Do you mean the daemon exited? Otherwise we'll have to find out who
>stopped the server; EXWM only stops the one it creates internally.


The daemon is still running. It was started with the custom name, and the corresponding socket can't be found after leaving EXWM with C-x C-c. This is reproducible. In my case, there is a Gentoo-specific wrapper script that starts and stops the daemon; I have to isolate its possible effects first. 


I believe all these are separate issues and should be treated as such. I'll create the corresponding Github issues. Regarding the current topic, I conclude the root cause to be smart-mode-line package but it'd be much better if somebody else confirmed this.
